### ISSUE TITLE
Mocking electron modules

### ISSUE THREAD

**Issue by davej** [electron/spectron#94]
Is there any way to mock electron modules from spectron?

Currently, I have to add all my mocks to the main process directly in my application code. So my mocks get shipped with the app, which is far from ideal.

My current setup looks something like the following:

`tests/e2e.js`
```js
// ...
new Application({
  path: electronPath,
  env: { RUNNING_IN_SPECTRON: '1' },
  args: ['.'],
});
// ...
```

`main.js`
```js
const electron = require('electron');
// ...
if (process.env.RUNNING_IN_SPECTRON) {
  electron.dialog.showOpenDialog = (opts, cb) => {
    if (opts.title === 'Choose Site Folder') {
      cb(['/Users/dave/lorem/ipsum/My Site']);
    }
  };
}
// ...
```

**Comment by kevinsawicki** [comment]
Yeah, this isn't currently possible since spectron only have access to run code in the renderer process and these APIs would need to be mocked in the main process.

One thing you could do now is add something like `['.', '-r', './tests/mocks.js']` to the `args` array and preload the mocks that way to no pollute your regular app code.

**Comment by davej** [comment]
> One thing you could do now is add something like ['.', '-r', './tests/mocks.js'] to the args array and preload the mocks that way to no pollute your regular app code.

Oh, great idea! Didn't think of that, I'll give it a go now.

**Comment by davej** [comment]
@kevinsawicki: Hmmm, didn't seem to work for me?

```js
    app = new Application({
      path: electronPath,
      env: { SPECTRON: '1' },
      args: ['.', '-r', join(__dirname, 'mocks.js')],
    });
```

**Comment by kevinsawicki** [comment]
>  args: ['.', '-r', join(__dirname, 'mocks.js')]

Sorry, I think `'.'` needs to be last in the array.

**Comment by davej** [comment]
@kevinsawicki 👍 Working a charm now!

**Comment by topmat** [comment]
This method was working on development build. 
  this.app = new Application({
    path: electron,
    args: ['-r', `${__dirname}/../mocks.js`, 'app']
  });


When I tried to test packed app. That doesn't working
  this.app = new Application({
    path: './release/mac/MyApp.app/Contents/MacOS/MyApp',
    args: ['-r', `${__dirname}/../mocks.js`, './release/mac/MyApp.app/Contents/Resources/app/main.js']
  });

Any idea on this?

**Comment by etiennejcharles** [comment]
@topmat
Given this file structure. 

![image](https://cloud.githubusercontent.com/assets/2872476/21724187/5d702d06-d401-11e6-88e1-9d6784d30373.png)

This is what I did to get it right. 

`args: [path.join(__dirname, '..'), '-r', path.join(__dirname, 'mocks.js')]`

1) The first argument correspond to the path of your **main.js** file. (wherever that may be)

2) The second arguments corresponds to the **require** statement to preload with `-r` your files (in your case, your mocks). i.e `electron main.js -r test/mocks.js ` should work in your console if you have electron installed globally. 

3) The 3rd corresponds to your **mocks path**.

**Comment by etiennejcharles** [comment]
@davej : Are you able to provide your mock.js file ? I can't seem to get it right either ~

**Comment by etiennejcharles** [comment]
Given this file structure.

![image](https://cloud.githubusercontent.com/assets/2872476/21724187/5d702d06-d401-11e6-88e1-9d6784d30373.png)

This is what I did to get it right.

`args: [path.join(__dirname, '..'), '-r', path.join(__dirname, 'mocks.js')]`

1) The first argument correspond to the path of your **main.js** file. (wherever that may be)

2) The second arguments corresponds to the **require** statement to preload with `-r` your files (in your case, your mocks). i.e `electron main.js -r test/mocks.js ` should work in your console if you have electron installed globally.

3) The 3rd corresponds to your **mocks path**. (edited)

In my main JS I did the following. 
```
...
const mock = require('./test/mock')
...
function createWindow () {
  if(process.env.RUNNING_IN_SPECTRON){
      mock(dialog);
  }
...
````

**Comment by Miw0** [comment]
@etiennejcharles is preloading "mock.js" really necessary? It seems that `const mock = require('./test/mock')` is enough.

Also this seems to be dirty, having mocks in your production code :neutral_face:

**Comment by cmdcolin** [comment]
The spectron-fake-dialog https://www.npmjs.com/package/spectron-fake-dialog seems like a good example of doing this mock, mentioned in another thread but reposting here :)

**Comment by zeke** [comment]
Here's how I recently mocked `require('electron')`:

https://github.com/electron/update-electron-app/blob/0dac0813245fe5e872a407ea61e661e8bd49588e/test.js#L7-L22

**Comment by ccorcos** [comment]
Hmm. This "-r" trick isnt working for me. Unfortunately, I have no idea where to find the cli args documentation either... :/

**Comment by jeremypatrickdahan** [comment]
@ccorcos The only place where I could find "documentation" was ... by typing `electron -h` in a terminal. I think this is actually a node option. Note that `electron -r mocks.js main.js` works but `electron main.js -r mocks.js` does not.
However, I can load the mock file but I did not manage to replace the electron modules.

**Comment by EmilyCamacho** [comment]
How are you all running your tests? Im using the following to run a jest/spectron test 

 `npm test filename `

I've tried OP's setup and @etiennejcharles but it doesn't seem to be working. What am I missing?

**Comment by etiennejcharles** [comment]
> How are you all running your tests? Im using the following to run a jest/spectron test
> 
> `npm test filename `
> 
> I've tried OP's setup and @etiennejcharles but it doesn't seem to be working. What am I missing?

Yeah it's been a while I've worked with Electron @EmilyCamacho  - the API's most likely have changed since I wrote the above - I don't even use the same editor 😆  

Wish there was a to be better way to do this 😅  - Sorry I can't help
